### DISCUSSION TITLE
Decoding Base64 json responses

### DISCUSSION THREAD

**Discussion by riteshnoronha** [Orange-OpenSource/hurl#3254] [Q&A] [answered]
Hi Everyone,

I believe currently there is no way to do this, but wanted to verify. 

if my response is a base64 encoded json, there is no way to decode this response and then assert using jsonpath correct?

Thanks
Ritesh

**Comment by lepapareil** [comment]
Hi @riteshnoronha, `base64Encode` and `base64Decode` are in our todo list => https://github.com/Orange-OpenSource/hurl/issues/2145

**Comment by lepapareil** [comment]
Hi @riteshnoronha,

As a workaround you can use a temporary base64 decoding api server.
Here is an example from scratch to do this hack, just creating a fake server giving encoded json body and another fake server decoding it:

- create a base64 encoded response script `/tmp/base64_encoded_api.sh`

```
#!/bin/bash
echo "HTTP/1.1 200 OK"
echo "Content-Type: text/plain"
echo "Connection: Closed"
echo
echo "eyJuYW1lIjogInRlc3QifQ==" 
```

- create a base64 header decoding response script `/tmp/base64_decode_api.sh`

```
#!/bin/bash
while read -r line ; do
    if [[ "${line}" =~ base64 ]] ; then
        cleaned=$(echo "${line}" | strings | cut --delimiter ':' --field 2 | tr -d ' ')
        decoded=$(echo "${cleaned}" | base64 --decode)
        echo "HTTP/1.1 200 OK"
        echo "Content-Type: application/json"
        echo "Connection: Closed"
        echo
        echo "${decoded}"
        break
    else
        continue
    fi
done
```

- start mini socat api servers executing scripts:

```
$ socat TCP-LISTEN:4321,reuseaddr,fork EXEC:/tmp/base64_encoded_api.sh &
$ socat TCP-LISTEN:1234,reuseaddr,fork EXEC:/tmp/base64_decode_api.sh &
```

- Create a Hurl file that gets base64 encoded json body and decodes it:

```
GET http://localhost:4321/get_base64_encoded_json_body
HTTP 200
[Captures]
encoded_body: body

GET http://localhost:1234/decode_base64_json_body_by_base64_header
base64: {{encoded_body}}
HTTP 200
[Captures]
decoded_body: body

[Asserts]
variable "decoded_body" jsonpath "$.name" == "test"
```

**Reply by riteshnoronha** [reply]
interesting, awesome will try this out. thanks @lepapareil

**Comment by jcamiel** [comment, marked as answer]
Just revisiting this issue, you can chain [filters](https://hurl.dev/docs/filters.html) and decodes from base 64 then chain with a JSONPath filter:

```
GET http://foo.com/toto.bin
HTTP 200
[Asserts]
body base64Decode decode "utf-8" jsonpath "$.name" == "toto"
```

The different "steps" of the chains are applied from left to right:
 
- `body` get HTTP response bytes and decodes it to a string
- `base64Decode` decodes the string to bytes
- `decode "utf-8"`: decodes bytes to string using "utf-8" encoding
- ` jsonpath "$.name"` apply a JSONPath query to a JSON


`base64Decode` is available from Hurl 6.1.0
